colcon build  make params -j4 doesn't work

I am using Qt creator 13.1, ros_qtc_plugin 13.  Ubuntu22.  ROS2 humble. 

When I run Build ROS2 project, my ubuntu 22 always stuck, my cpu has 24 kernel, I checked ROS log, it used -j24 -L24,  I want to reduce number of thread then Build ROS2 using colcon build.

In Qt creator - Build & Run - Colcon detail, I can find Colcon Build / CMake / Make parameterinput. I added "--parallel-workers=1" in Colcon Build params input, it has no error. But when I add "-j 4" in Make parameter, and Build in Qt Creator, it fails,  "can not find -j in CMake....."

![Screenshot from 2024-07-22 09-04-56](https://github.com/user-attachments/assets/3c33acab-02d4-4ad1-8634-580bf5a9e6f9)
I haved added -j4 param in Qt creator, and when I build, it reports error:
CMake Error: Unknown argument -j4

I put -j4 in Make argument.


09:04:31: Running steps for project btl_ros_project...
09:04:31: Starting: "/usr/bin/colcon" build --parallel-workers 1 --event-handlers status+ console_start_end+ --cmake-args -G 'CodeBlocks - Unix Makefiles' -DCMAKE_BUILD_TYPE=Debug --make-args -j4
Starting >>> beta_interfaces
--- stderr: beta_interfaces
CMake Error: Unknown argument -j4
CMake Error: Run 'cmake --help' for all supported options.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/colcon_core/executor/__init__.py", line 91, in __call__
    rc = await self.task(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/colcon_core/task/__init__.py", line 93, in __call__
    return await task_method(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/colcon_ros/task/ament_cmake/build.py", line 59, in build
    rc = await extension.build(
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/build.py", line 87, in build
    rc = await self._reconfigure(args, env)
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/build.py", line 182, in _reconfigure
    buildfile = get_buildfile(cmake_cache)
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/__init__.py", line 114, in get_buildfile
    if 'Visual Studio' in generator:
TypeError: argument of type 'NoneType' is not iterable
---
Failed   <<< beta_interfaces [0.01s, exited with code 1]
                     
Summary: 0 packages finished [0.12s]
  1 package failed: beta_interfaces
  1 package had stderr output: beta_interfaces
  2 packages not processed
argument of type 'NoneType' is not iterable
09:04:31: The process "/usr/bin/colcon" exited with code 1.
Error while building/deploying project btl_ros_project (kit: Desktop)
When executing step "Colcon Step"
09:04:31: Elapsed time: 00:00.

![Screenshot from 2024-07-22 09-04-56](https://github.com/user-attachments/assets/3c33acab-02d4-4ad1-8634-580bf5a9e6f9)
I haved added -j4 param in Qt creator, and when I build, it reports error:
CMake Error: Unknown argument -j4

I put -j4 in Make argument.


09:04:31: Running steps for project btl_ros_project...
09:04:31: Starting: "/usr/bin/colcon" build --parallel-workers 1 --event-handlers status+ console_start_end+ --cmake-args -G 'CodeBlocks - Unix Makefiles' -DCMAKE_BUILD_TYPE=Debug --make-args -j4
Starting >>> beta_interfaces
--- stderr: beta_interfaces
CMake Error: Unknown argument -j4
CMake Error: Run 'cmake --help' for all supported options.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/colcon_core/executor/__init__.py", line 91, in __call__
    rc = await self.task(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/colcon_core/task/__init__.py", line 93, in __call__
    return await task_method(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/colcon_ros/task/ament_cmake/build.py", line 59, in build
    rc = await extension.build(
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/build.py", line 87, in build
    rc = await self._reconfigure(args, env)
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/build.py", line 182, in _reconfigure
    buildfile = get_buildfile(cmake_cache)
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/__init__.py", line 114, in get_buildfile
    if 'Visual Studio' in generator:
TypeError: argument of type 'NoneType' is not iterable
---
Failed   <<< beta_interfaces [0.01s, exited with code 1]
                     
Summary: 0 packages finished [0.12s]
  1 package failed: beta_interfaces
  1 package had stderr output: beta_interfaces
  2 packages not processed
argument of type 'NoneType' is not iterable
09:04:31: The process "/usr/bin/colcon" exited with code 1.
Error while building/deploying project btl_ros_project (kit: Desktop)
When executing step "Colcon Step"
09:04:31: Elapsed time: 00:00.

screenshot: 
![Screenshot from 2024-07-22 09-04-56](https://github.com/user-attachments/assets/0b23239b-1201-4840-b90b-018dc69efb0c)


I haved added -j4 param in Qt creator, and when I build
![Screenshot from 2024-07-22 09-04-56](https://github.com/user-attachments/assets/6af12ca5-6004-4e52-ba00-b4de025b50b3)
, it reports error:
CMake Error: Unknown argument -j4

I put -j4 in Make argument.


09:04:31: Running steps for project btl_ros_project...
09:04:31: Starting: "/usr/bin/colcon" build --parallel-workers 1 --event-handlers status+ console_start_end+ --cmake-args -G 'CodeBlocks - Unix Makefiles' -DCMAKE_BUILD_TYPE=Debug --make-args -j4
Starting >>> beta_interfaces
--- stderr: beta_interfaces
CMake Error: Unknown argument -j4
CMake Error: Run 'cmake --help' for all supported options.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/colcon_core/executor/__init__.py", line 91, in __call__
    rc = await self.task(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/colcon_core/task/__init__.py", line 93, in __call__
    return await task_method(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/colcon_ros/task/ament_cmake/build.py", line 59, in build
    rc = await extension.build(
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/build.py", line 87, in build
    rc = await self._reconfigure(args, env)
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/build.py", line 182, in _reconfigure
    buildfile = get_buildfile(cmake_cache)
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/__init__.py", line 114, in get_buildfile
    if 'Visual Studio' in generator:
TypeError: argument of type 'NoneType' is not iterable
---
Failed   <<< beta_interfaces [0.01s, exited with code 1]
                     
Summary: 0 packages finished [0.12s]
  1 package failed: beta_interfaces
  1 package had stderr output: beta_interfaces
  2 packages not processed
argument of type 'NoneType' is not iterable
09:04:31: The process "/usr/bin/colcon" exited with code 1.
Error while building/deploying project btl_ros_project (kit: Desktop)
When executing step "Colcon Step"
09:04:31: Elapsed time: 00:00.

> Does the generated command line `/usr/bin/colcon build --parallel-workers 1 --event-handlers status+ console_start_end+ --cmake-args -G 'CodeBlocks - Unix Makefiles' -DCMAKE_BUILD_TYPE=Debug --make-args -j4` work for you in the terminal outside of Qt Creator?

I run the command out of Qt, and it doesn't work for me. 

```
(base) coco@coco-YangTianT4900k-93IRB:~/Downloads$ /usr/bin/colcon build --parallel-workers 1 --event-handlers status+ console_start_end+ --cmake-args -G 'CodeBlocks - Unix Makefiles' -DCMAKE_BUILD_TYPE=Debug --make-args -j4
Starting >>> boost_core
--- stderr: boost_core
CMake Error: Unknown argument -j4
CMake Error: Run 'cmake --help' for all supported options.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/colcon_core/executor/__init__.py", line 91, in __call__
    rc = await self.task(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/colcon_core/task/__init__.py", line 93, in __call__
    return await task_method(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/build.py", line 87, in build
    rc = await self._reconfigure(args, env)
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/build.py", line 182, in _reconfigure
    buildfile = get_buildfile(cmake_cache)
  File "/usr/lib/python3/dist-packages/colcon_cmake/task/cmake/__init__.py", line 114, in get_buildfile
    if 'Visual Studio' in generator:
TypeError: argument of type 'NoneType' is not iterable
---
Failed   <<< boost_core [0.03s, exited with code 1]

Summary: 0 packages finished [5.68s]
  1 package failed: boost_core
  1 package had stderr output: boost_core
  193 packages not processed
```

--- a/src/project_manager/ros_colcon_step.cpp
+++ b/src/project_manager/ros_colcon_step.cpp
@@ -51,7 +51,6 @@ const char ROS_COLCON_STEP_DISPLAY_NAME[] = QT_TRANSLATE_NOOP("ROSProjectManager
 const char ROS_COLCON_STEP[] = "ROSProjectManager.ROSColconStep.Target";
 const char ROS_COLCON_STEP_ARGUMENTS_KEY[] = "ROSProjectManager.ROSColconStep.ColconArguments";
 const char ROS_COLCON_STEP_CMAKE_ARGUMENTS_KEY[] = "ROSProjectManager.ROSColconStep.CMakeArguments";
-const char ROS_COLCON_STEP_MAKE_ARGUMENTS_KEY[] = "ROSProjectManager.ROSColconStep.MakeArguments";
 
 ROSColconStep::ROSColconStep(BuildStepList *parent, const Utils::Id id) :
     AbstractProcessStep(parent, id)
@@ -147,7 +146,6 @@ void ROSColconStep::toMap(Utils::Store &map) const
     map.insert(ROS_COLCON_STEP, m_target);
     map.insert(ROS_COLCON_STEP_ARGUMENTS_KEY, m_colconArguments);
     map.insert(ROS_COLCON_STEP_CMAKE_ARGUMENTS_KEY, m_cmakeArguments);
-    map.insert(ROS_COLCON_STEP_MAKE_ARGUMENTS_KEY, m_makeArguments);
 }
 
 void ROSColconStep::fromMap(const Utils::Store &map)
@@ -155,7 +153,6 @@ void ROSColconStep::fromMap(const Utils::Store &map)
     m_target = (BuildTargets)map.value(ROS_COLCON_STEP).toInt();
     m_colconArguments = map.value(ROS_COLCON_STEP_ARGUMENTS_KEY).toString();
     m_cmakeArguments = map.value(ROS_COLCON_STEP_CMAKE_ARGUMENTS_KEY).toString();
-    m_makeArguments = map.value(ROS_COLCON_STEP_MAKE_ARGUMENTS_KEY).toString();
 
     BuildStep::fromMap(map);
 }
@@ -189,9 +186,6 @@ QString ROSColconStep::allArguments(ROSUtils::BuildType buildType, bool includeD
 
         break;
     }
-
-    if (!m_makeArguments.isEmpty())
-        args << QString("--make-args %1").arg(m_makeArguments);
 
     return args.join(" ");
 }
@@ -250,7 +244,6 @@ ROSColconStepWidget::ROSColconStepWidget(ROSColconStep *makeStep)
 
     m_ui->colconArgumentsLineEdit->setText(m_makeStep->m_colconArguments);
     m_ui->cmakeArgumentsLineEdit->setText(m_makeStep->m_cmakeArguments);
-    m_ui->makeArgumentsLineEdit->setText(m_makeStep->m_makeArguments);
 
     updateDetails();
 
@@ -258,9 +251,6 @@ ROSColconStepWidget::ROSColconStepWidget(ROSColconStep *makeStep)
             this, &ROSColconStepWidget::updateDetails);
 
     connect(m_ui->cmakeArgumentsLineEdit, &QLineEdit::textEdited,
-            this, &ROSColconStepWidget::updateDetails);
-
-    connect(m_ui->makeArgumentsLineEdit, &QLineEdit::textEdited,
             this, &ROSColconStepWidget::updateDetails);
 
     connect(m_makeStep, SIGNAL(enabledChanged()),
@@ -294,7 +284,6 @@ void ROSColconStepWidget::updateDetails()
 {
     m_makeStep->m_colconArguments = m_ui->colconArgumentsLineEdit->text();
     m_makeStep->m_cmakeArguments = m_ui->cmakeArgumentsLineEdit->text();
-    m_makeStep->m_makeArguments = m_ui->makeArgumentsLineEdit->text();
 
     ROSBuildConfiguration *bc = m_makeStep->rosBuildConfiguration();
     ROSUtils::WorkspaceInfo workspaceInfo = ROSUtils::getWorkspaceInfo(bc->project()->projectDirectory(), bc->rosBuildSystem(), bc->project()->distribution());

--- a/src/project_manager/ros_colcon_step.h
+++ b/src/project_manager/ros_colcon_step.h
@@ -74,7 +74,6 @@ private:
     BuildTargets m_target;
     QString m_colconArguments;
     QString m_cmakeArguments;
-    QString m_makeArguments;
     QRegularExpression m_percentProgress;
 };
 

--- a/src/project_manager/ros_colcon_step.ui
+++ b/src/project_manager/ros_colcon_step.ui
@@ -7,12 +7,12 @@
     <x>0</x>
     <y>0</y>
     <width>545</width>
-    <height>105</height>
+    <height>79</height>
    </rect>
   </property>
   <layout class="QFormLayout" name="formLayout">
    <property name="fieldGrowthPolicy">
-    <enum>QFormLayout::ExpandingFieldsGrow</enum>
+    <enum>QFormLayout::FieldGrowthPolicy::ExpandingFieldsGrow</enum>
    </property>
    <property name="leftMargin">
     <number>9</number>
@@ -52,22 +52,6 @@
    </item>
    <item row="1" column="1">
     <widget class="QLineEdit" name="cmakeArgumentsLineEdit"/>
-   </item>
-   <item row="2" column="0">
-    <widget class="QLabel" name="makeArgumentsLabel">
-     <property name="sizePolicy">
-      <sizepolicy hsizetype="Preferred" vsizetype="Preferred">
-       <horstretch>0</horstretch>
-       <verstretch>0</verstretch>
-      </sizepolicy>
-     </property>
-     <property name="text">
-      <string>Make Arguments:</string>
-     </property>
-    </widget>
-   </item>
-   <item row="2" column="1">
-    <widget class="QLineEdit" name="makeArgumentsLineEdit"/>
    </item>
   </layout>
  </widget>

